Better download_and_gen script, support for perfmon-metrics

captain5050 opened the pull request:
Hi,

these changes refactor existing converter scripts to be usable as python modules. A new converter script is added written in python. The new script automates the creation of the perf mapfile.csv highlighting some errors in upstream Linux. The new script downloads files using the descriptions in mapfile.csv, this means version changes don't need to be made to the script. The script uses Python's urllib library to avoid downloading then processing files, processing is done on the in-memory version. Perfmon-metrics events are added except in a few cases where errors were detected with 'perf test' in the github version and because of the known #SYSTEM_TSC_FREQ issue. Once errors in perfmon-metrics are fixed it is straightforward to enable generation of the metrics again.

The output of the new script matches the previous download_and_gen.sh except, the addition of mapfile.csv and perfmon-metrics, and uncore events are generated for architectures where no csv file is present  (by passing an empty csv file). As such the output is the same or superior.

Thanks,
Ian

captain5050 committed:
Add meteorlake and update 01.org json file versions

Meteorlake is available on 01.org but very brief, missing floating point
for example. Add to the download_and_gen.sh along with update to
existing architecture versions.

captain5050 committed:
Make hybrid-json-to-perf-json.py into a module

Allow it to be used by other python programs as an import.

captain5050 committed:
Modularize uncore_csv_json

Minor tweaks to arg parsing to better parse file arguments.

captain5050 committed:
Modularie extract-tma-metrics

Allow extract-tma-metrics to be used as a module by other python
programs. Function layout changes necessary to achieve this. Run
through pyformat.

captain5050 committed:
Add output option to extract-tma-metrics

Default to output to stdout but allow a command line override.

captain5050 committed:
Move print output to be behind verbose flag

This reduces unecessary output when processing.

captain5050 committed:
Fix generation of uncore experimental events

Fix a typo that meant uncore experimental events were being ignored.

captain5050 committed:
Fix TGL columns for BDW-DE

BDW-DE no longer exists in TMA metrics, so fix the column headings
accordingly.

captain5050 committed:
Add helper to get TMA column header name

TMA metrics columns may combine multiple architectures with a / like
ADL/RPL. Use the tables in extract-tma-metrics to help do this.
BDW-DE was removed from TMA metrics, hardware its equivalence to BDX.

captain5050 committed:
Rewrite download_and_gen.sh in python

Use mapfile.csv for the version numbers and directly download the files
while generating perf json, rather than downloading then working on
files.
Add an ability to generate a perf mapfile.csv.

captain5050 committed:
Add metrics from intel/perfmon-metrics

When reading the mapfile.csv, see if a perfmon-metrics directory exists.
If it does then add these metrics to the appropriate architecture. As
some of the metrics fail on upstream perf they are dropped, see the data
structure 'broken_extra_metrics'. If duplicate metrics exist then the
perfmon-metrics version is preferred.

Tested on BDX, CLX, ICX, SKX, SPR.

captain5050 committed:
Add missing SPR to TMA metric generation

captain5050 committed:
Add verbose flag to uncore_csv_json

This makes it easier to distinguish proper errors from warnings.

captain5050 committed:
Remove download_and_gen.sh

Prefer download_and_gen.py that doesn't require updates for version
numbers.

ZhengjunXing0 commented on extract-tma-metrics.py:
```diff
@@ -33,500 +33,604 @@
+            'IVB', 'JKT/SNB-EP', 'SNB'),
+    'GRT': ('GRT'),
 }
 
```
The old one looks better, why change this?

captain5050 replied:
Generally at Google we use automatic code formatting, it avoids arguments over white space and lets you focus on the substance of the change. The change here was caused by pyformat.

ZhengjunXing0 commented on extract-tma-metrics.py:
```diff
@@ -33,500 +33,604 @@
+    (['KNL', 'KNM'], [6], [2, 3, 6]),
+    (['GLM', 'SNR'], [1, 3, 6], [2, 3, 6, 10]),
 ]
 
```
The old one looks better, why change this?

captain5050 replied:
pyformat. We can hand indent but there were issues with line length, etc. with the old code.

captain5050 committed:
Ensure encoding is always done to ASCII

S390 and Power Linux perf json files contain non-ASCII characters.
Ensure this doesn't happen here.

captain5050 committed:
Add hermetic-download command line option

Option to download all the relevant csv and json files as computed by
the parsing of 01.org's mapfile.csv. This allows the script to run
without any network access. When downloading the files fix non-ascii
characters like (R) and (TM).

ZhengjunXing0 commented:
Applied, thanks.
